sdks/python/apache_beam/io/gcp/dicomio.py

```diff
@@ -0,0 +1,419 @@
+    Key-value pairs:
+      project_id: Id of the project in which DICOM store locates. (Required)
+      region: Region where the DICOM store resides. (Required)
+      dataset_id: Id of the dataset where DICOM store belongs to. (Required)
+      dicom_store_id: Id of the dicom store. (Required)
+      search_type: Which type of search it is, could only be one of the three
+        values: 'instances', 'series', or 'studies'. (Required)
+      params: A dict of str:str pairs used to refine QIDO search. (Optional)
+        Supported tags in three categories:
+          1. Studies:
+            StudyInstanceUID
+            PatientName
+            PatientID
+            AccessionNumber
+            ReferringPhysicianName
+            StudyDate
+          2. Series: all study level search terms and
+            SeriesInstanceUID
+            Modality
+          3. Instances: all study/series level search terms and
+            SOPInstanceUID
+        e.g. {"StudyInstanceUID":"1","SeriesInstanceUID":"2"}
+    OUTPUT:
+    The output dict wraps results as well as error messages:
+    {
+      'result': a list of dicts in JSON style.
+      'success': boolean value telling whether the operation is successful.
+      'input': detail ids and dicomweb path for this retrieval.
+      'status': status code from the server, used as error message.
+    }
+  """
+  def __init__(self, credential=None):
+    """Initializes DicomSearch.
+    Args:
+      credential: # type: Google credential object, if it is specified, the
+        Http client will use it to create sessions instead of the default.
+    """
+    self.credential = credential
+
+  def expand(self, pcoll):
+    return pcoll | beam.ParDo(_QidoSource(self.credential))
+
+
+class _QidoSource(beam.DoFn):
+  """A DoFn for executing every qido query request."""
+  def __init__(self, credential=None):
+    self.credential = credential
+
+  def process(self, element):
+    # Check if all required keys present.
+    required_keys = [
+        'project_id', 'region', 'dataset_id', 'dicom_store_id', 'search_type'
+    ]
+
+    error_message = None
+
+    for key in required_keys:
+      if key not in element:
+        error_message = 'Must have %s in the dict.' % (key)
+        break
+
+    if not error_message:
+      project_id = element['project_id']
+      region = element['region']
+      dataset_id = element['dataset_id']
+      dicom_store_id = element['dicom_store_id']
+      search_type = element['search_type']
+      params = element['params'] if 'params' in element else None
+
+      # Call qido search http client
+      if element['search_type'] in ['instances', "studies", "series"]:
+        result, status_code = DicomApiHttpClient().qido_search(
+          project_id, region, dataset_id, dicom_store_id,
+          search_type, params, self.credential
+        )
+      else:
+        error_message = (
+            'Search type can only be "studies", '
+            '"instances" or "series"')
```

Comment:
Performing these API calls is blocking. Because we're only issuing one call per thread, the workers will be blocked waiting for IO, and CPU will be wasted. You can improve this with a thread pool. You would do something like this:

```
class MyDoFn(DoFn):
  def start_bundle(self):
    self.buffer = []

  def process(self, element, window=DoFn.WindowParam):
    if validate_element(element):
      self.buffer.append((element, window))
    else:
      yield build_output_with_error_message(element)

  def _make_dicom_http_request(self, element):
    # Element is validated.
    result, status = DicomApiHttpClient().quido_search(.....)
    ....

  def finish_bundle(self):
    thread_pool = ThreadPoolExecutor()
    futures = [thread_pool.submit(_make_dicom_http_request, elm)  # Something like this
                   for elm in self.buffer]
    # Note that here you have to yield a WindowedElement, not just the plain element
    # This is because one bundle can have elements belonging to multiple windows.
    yield [f.get() for f in futures]
```

This is certainly not a blocker for the PR, but users will hit this issue 100% for sure, so it's worth doing now. Feel free to add a TODO (and get acknowledgement from your mentor), or to implement it.

Reply:
I'm also happy to clarify further on what the code snippet means if you'd like.

Reply:
This is really interesting one, I will give it a try!

Reply:
ah one more thing. You should make sure that you're not buffering too much (in streaming, bundles are usually many and small, but in batch bundles can contain thousands of elements. I'd recommend setting a maximum buffer size, and flushing if it's reached)

Reply:
@pabloem The batch processing is implemented. However I am not sure if my way of setting timestamp is correct, please double check :)